File: src/api/routes/pedidos.py
from fastapi import APIRouter, responses
from pydantic import BaseModel
from datetime import date, datetime
from api.services import database_pedidos, respostas_padrao

router = APIRouter()

class BasePedido(BaseModel):
    cliente_id: int
    status: str = "PENDENTE"
    data_entrega: date | None = None

class ItemPedidoInput(BaseModel):
    formula_id: int
    quantidade: int

@router.get("/")
async def listar_pedidos():
    resultado = database_pedidos.listar_pedidos()
    return responses.JSONResponse(content=resultado, status_code=200)

@router.get("/{id}")
async def listar_pedido_id(id: int):
    resultado = database_pedidos.listar_pedido_id(id)
    if resultado is None:
        return respostas_padrao.nao_encontrado
    return responses.JSONResponse(content=resultado, status_code=200)

@router.post("/")
async def criar_pedido(body: BasePedido):
    pedido_id = database_pedidos.criar_pedido(
        cliente_id=body.cliente_id,
        status=body.status,
        data_entrega=body.data_entrega
    )
    return responses.JSONResponse(
        content={"id": pedido_id, "status": "Ok"}, status_code=200
    )

@router.put("/{id}")
async def atualizar_pedido(id: int, body: BasePedido):
    resultado = database_pedidos.editar_pedido(
        id=id,
        cliente_id=body.cliente_id,
        status=body.status,
        data_entrega=str(body.data_entrega) if body.data_entrega else None
    )
    return responses.PlainTextResponse(content=resultado, status_code=200)

@router.delete("/{id}")
async def deletar_pedido(id: int):
    resultado = database_pedidos.deletar_pedido(id)
    return responses.PlainTextResponse(content=resultado, status_code=200)

@router.get("/{id}/itens")
async def listar_itens_pedido(id: int):
    resultado = database_pedidos.listar_itens_pedido(id)
    return responses.JSONResponse(content=resultado, status_code=200)

@router.post("/{id}/itens")
async def adicionar_item_pedido(id: int, body: ItemPedidoInput):
    resultado = database_pedidos.adicionar_item_pedido(
        pedido_id=id,
        formula_id=body.formula_id,
        quantidade=body.quantidade
    )
    return responses.PlainTextResponse(content=resultado, status_code=200)

File: src/web/app/js/pages/pedidos.js
async function renderPedidos() {
    const res = await API.listarPedidos();
    const data = res.ok ? res.data : [];
    let html = `<div class="d-flex justify-content-between mb-3">
        <p></p>
        <button class="btn btn-primary btn-sm" onclick="pedidoForm(null)"><i class="bi bi-plus-lg"></i> Novo</button>
    </div>`;
    html += renderTable(
        ["ID", "Cliente ID", "Data Pedido", "Status", "Data Entrega"],
        data.map(p => [p.id, p.cliente_id, p.data_pedido,
            `<span class="badge ${statusBadge(p.status)}">${p.status}</span>`,
            p.data_entrega || "-"]),
        r => `<button class="btn btn-sm btn-outline-info me-1" onclick="pedidoForm(${r[0]})"><i class="bi bi-pencil"></i></button>
              <button class="btn btn-sm btn-outline-warning me-1" onclick="verItensPedido(${r[0]})"><i class="bi bi-list-ul"></i></button>
              <button class="btn btn-sm btn-outline-danger" onclick="pedidoDelete(${r[0]})"><i class="bi bi-trash"></i></button>`
    );
    document.getElementById("content-body").innerHTML = html;
}

window.pedidoForm = async function (id) {
    let p = { cliente_id: "", status: "PENDENTE", data_entrega: "" };
    if (id) { const r = await API.buscarPedido(id); if (r.ok && r.data) p = r.data; }
    showModal(id ? "Editar Pedido" : "Novo Pedido",
        formGroup("Cliente ID", "f-cli", "number", p.cliente_id) +
        selGroup("Status", "f-status", [
            { value: "PENDENTE", label: "Pendente" }, { value: "EM_PRODUCAO", label: "Em Producao" },
            { value: "FINALIZADO", label: "Finalizado" }, { value: "CANCELADO", label: "Cancelado" }
        ], p.status) +
        formGroup("Data Entrega", "f-entrega", "date", p.data_entrega || ""),
        async function () {
            const d = { cliente_id: Number(val("f-cli")), status: val("f-status"), data_entrega: val("f-entrega") || null };
            if (id) await API.atualizarPedido(id, d); else {
                const cr = await API.criarPedido(d);
                if (cr.ok && cr.data && cr.data.id) {
                    closeModal();
                    await adicionarItensPedido(cr.data.id);
                    return;
                }
            }
            closeModal(); renderPedidos();
        }
    );
};

window.verItensPedido = async function (id) {
    const res = await API.buscarPedido(id);
    const pedido = res.ok ? res.data : {};
    const itens = pedido.itens || [];
    let rows = itens.map(i => `<tr><td>${i.id}</td><td>${i.formula_id}</td><td>${i.quantidade}</td></tr>`).join("");
    showModal("Itens do Pedido #" + id,
        `<button class="btn btn-sm btn-primary mb-2" onclick="adicionarItensPedido(${id})"><i class="bi bi-plus-lg"></i> Adicionar</button>
        <div class="table-wrap"><table class="table table-dark table-sm">
        <thead><tr><th>ID</th><th>Formula ID</th><th>Quantidade</th></tr></thead>
        <tbody>${rows || '<tr><td colspan="3" class="text-center text-muted">Nenhum item</td></tr>'}</tbody></table></div>`,
        null
    );
};

window.adicionarItensPedido = async function (pedidoId) {
    showModal("Adicionar Item ao Pedido",
        formGroup("Formula ID", "f-fml", "number", "") +
        formGroup("Quantidade", "f-qtd", "number", ""),
        async function () {
            const d = { formula_id: Number(val("f-fml")), quantidade: Number(val("f-qtd")) };
            await API.adicionarItemPedido(pedidoId, d);
            if (confirm("Item adicionado! Adicionar mais?")) adicionarItensPedido(pedidoId);
            else { closeModal(); renderPedidos(); }
        }
    );
};

window.pedidoDelete = async function (id) { if (confirm("Deletar pedido?")) { await API.deletarPedido(id); renderPedidos(); } };
